### ISSUE TITLE
Discrepancy between HMC output in R and C++

### ISSUE THREAD

**Issue by LongPham7** [GeomScale/volesti#260]
I wrote two programs for HMC sampling: one using volesti's R interface and the other using volesti's C++ interface. Both programs sample from the same convex polytope and have the same parameters (e.g. Lipschitz constant and step size for HMC). However, the R code runs successfully, while the C++ code returns nonsensical results. Because my application uses C++, I need the C++ code to work well. I'd be grateful if someone could tell me what is going on. Thanks a lot in advance!

Concretely, I want to sample uniformly at random from a two-dimensional convex polytope defined by
```
0 <= x <= 5
0 <= y <= 5
-0.01 <= x - y <= 0.01
```
Here is my R code:
```R
library("volesti")

A <- matrix(rep(0, 6*2), nrow=6, ncol=2)
A[1,1] <- -1 # 0 <= x
A[2,1] <- 1 # x <= 5
A[3,2] <- -1 # 0 <= y
A[4,2] <- 1 # y <= 5
A[5,1] <- 1; A[5,2] <- -1 # x - y <= 0.01
A[6,1] <- -1; A[6,2] <- 1 # -x + y <= 0.01
b <- c(0, 5, 0, 5, 0.01, 0.01)
P <- Hpolytope$new(A, b)

# Smoothness and strong-convexity
L <- 4
m <- 4
starting_point <- c(0.9, 0.9)
cat("Starting point: ", warm_start, "\n")

n_samples <- 10
n_burns <- n_samples / 2
f <- function(x) {return(0)} # Netgative logprob of uniform distribution with density 1
grad_f <- function(x) {return(c(0,0))} # Negative logprob gradient of uniform distribution with density 1

parameters <- sample_points(P, n = n_samples, 
                            random_walk = list("walk" = "HMC", "step_size" = 1, "nburns" = n_burns, 
                                            "walk_length" = 150, "solver" = "leapfrog", "starting_point" = starting_point), 
                            distribution = list("density" = "logconcave", "negative_logprob" = f, "negative_logprob_gradient" = grad_f, "L_" = L, "m" = m))

for (i in 1:ncol(parameters)) {
    cat("Sample: ", parameters[, i], "\n")
}
```
It produces output like
```
Starting point:  0.9 0.9 
Sample:  4.912857 4.920996 
Sample:  0.387464 0.3902125 
Sample:  4.188567 4.197291 
Sample:  0.0495364 0.04044371 
Sample:  3.361872 3.358376 
Sample:  4.327392 4.336381 
Sample:  2.777365 2.783168 
Sample:  1.723236 1.726752 
Sample:  4.479537 4.489249 
Sample:  0.4050331 0.4031142 
```
This is consistent with what I expect. 

By contrast, my C++ code, which is a minor modification of the file `volesti/examples/logconcave/simple_hmc.cpp` in the volesti's project folder, is
```c++
#include <iostream>
#include <cmath>
#include <functional>
#include <vector>
#include <unistd.h>
#include <string>
#include <typeinfo>

#include "Eigen/Eigen"

#include "ode_solvers/ode_solvers.hpp"
#include "random.hpp"
#include "random/uniform_int.hpp"
#include "random/normal_distribution.hpp"
#include "random/uniform_real_distribution.hpp"
#include "random_walks/random_walks.hpp"
#include "volume/volume_sequence_of_balls.hpp"
#include "volume/volume_cooling_gaussians.hpp"
#include "volume/volume_cooling_balls.hpp"
#include "generators/known_polytope_generators.h"

struct CustomFunctor
{

    // Custom density with neg log prob equal to 0
    template <
        typename NT>
    struct parameters
    {
        unsigned int order;
        NT L;     // Lipschitz constant for gradient
        NT m;     // Strong convexity constant
        NT kappa; // Condition number

        parameters() : order(2), L(4), m(4), kappa(1){};
    };

    template <
        typename Point>
    struct GradientFunctor
    {
        typedef typename Point::FT NT;
        typedef std::vector<Point> pts;

        parameters<NT> &params;

        GradientFunctor(parameters<NT> &params_) : params(params_){};

        // The index i represents the state vector index
        Point operator()(unsigned int const &i, pts const &xs, NT const &t) const
        {
            if (i == params.order - 1)
                return 0.0 * Point::all_ones(xs[0].dimension());
            else
                return xs[i + 1]; // returns velocity
        }
    };

    template <
        typename Point>
    struct FunctionFunctor
    {
        typedef typename Point::FT NT;

        parameters<NT> &params;

        FunctionFunctor(parameters<NT> &params_) : params(params_){};

        // The index i represents the state vector index
        NT operator()(Point const &x) const
        {
            return 0.0;
        }
    };
};

template <typename NT>
void run_main()
{
    typedef Cartesian<NT> Kernel;
    typedef typename Kernel::Point Point;
    typedef HPolytope<Point> Hpolytope;
    typedef BoostRandomNumberGenerator<boost::mt19937, NT> RandomNumberGenerator;

    RandomNumberGenerator rng(1);
    unsigned int dim = 2;

    // Define a polytope
    typedef typename Hpolytope::MT MT;
    typedef typename Hpolytope::VT VT;
    MT A;
    VT b;
    A.resize(6, dim);
    b.resize(6);
    A(0, 0) = -1; A(0, 1) = 0; // 0 <= x
    A(1, 0) = 1; A(1, 1) = 0; // x <= 5
    A(2, 0) = 0; A(2, 1) = -1; // 0 <= y
    A(3, 0) = 0; A(3, 1) = 1; // y <= 5
    A(4, 0) = 1; A(4, 1) = -1; // x y <= 0.01
    A(5, 0) = -1; A(5, 1) = 1; // -x + y <= 0.01
    b(0) = 0; b(1) = 5; b(2) = 0; b(3) = 5; b(4) = 0.01; b(5) = 0.01;

    Hpolytope P{dim, A, b};
    std::cout << "Here is polytope P:";
    P.print();

    // Define functors
    typedef CustomFunctor::GradientFunctor<Point> NegativeGradientFunctor;
    typedef CustomFunctor::FunctionFunctor<Point> NegativeLogprobFunctor;
    CustomFunctor::parameters<NT> params;
    params.kappa = 2;

    NegativeGradientFunctor F(params);
    NegativeLogprobFunctor f(params);

    // Define a starting point
    Point x0 = 0.9 * Point::all_ones(dim);
    std::cout << "Starting point: " << x0.getCoefficients().transpose() << std::endl;

    // HMC walk
    HamiltonianMonteCarloWalk::parameters<NT, NegativeGradientFunctor> hmc_params(F, dim);

    // In the first argument put in the address of an H-Polytope
    // for truncated sampling and NULL for untruncated
    typedef LeapfrogODESolver<Point, NT, Hpolytope, NegativeGradientFunctor> Solver;
    HamiltonianMonteCarloWalk::Walk<Point, Hpolytope, RandomNumberGenerator, NegativeGradientFunctor, NegativeLogprobFunctor, Solver>
        hmc(&P, x0, F, f, hmc_params);

    hmc.disable_adaptive();
    hmc.solver->eta = 1; // step size
    unsigned int hmc_walk_length = 150;
    int n_samples = 10; // Half will be burned

    std::cout << "Reflective HMC result:\n";

    for (int i = 0; i < n_samples; i++)
    {
        hmc.apply(rng, hmc_walk_length);
        if (i > n_samples / 2)
            std::cout << hmc.x.getCoefficients().transpose() << std::endl;
    }

    std::cerr << "Step size (final): " << hmc.solver->eta << std::endl;
    std::cerr << "Discard Ratio: " << hmc.discard_ratio << std::endl;
    std::cerr << "Average Acceptance Log-prob: " << exp(hmc.average_acceptance_log_prob) << std::endl;
}

int main()
{
    run_main<double>();
    return 0;
}
```
It produces output like
```
Here is polytope P: 6 2 double
-1 0 <= 0
1 0 <= 5
0 -1 <= 0
0 1 <= 5
1 -1 <= 0.01
-1 1 <= 0.01
Starting point: 0.9 0.9
Reflective HMC result:
-nan  inf
-nan  inf
-nan  inf
-nan  inf
Step size (final): 1
Discard Ratio: 0
Average Acceptance Log-prob: 1
```
Thus, HMC in the C++ interface returns `-nan inf`, which is definitely not what I expected.

**Comment by vissarion** [comment]
Thanks for this report. @papachristoumarios will you have some time looking at it?

**Comment by papachristoumarios** [comment]
Hi @LongPham7, and thank you for letting us know. I think that there is a discrepancy between your C++ code and the R code. Could you please change the GradientFunctor to 

```
    template <
        typename Point>
    struct GradientFunctor
    {
        typedef typename Point::FT NT;
        typedef std::vector<Point> pts;

        parameters<NT> &params;

        GradientFunctor(parameters<NT> &params_) : params(params_){};

        Point operator()(unsigned int const &i, pts const &xs, NT const &t) const
        {
                return 0.0 * Point::all_ones(xs[0].dimension());
        }
    };
```

and let us know if it works? The rest of the code looks similar after a quick look.

**Comment by LongPham7** [comment]
Hi @vissarion and @papachristoumarios. Thanks a lot for your swift response. I modified `GradientFunctor` as you suggested, and here is the output I got:
```
Here is polytope P: 6 2 double
-1 0 <= 0
1 0 <= 5
0 -1 <= 0
0 1 <= 5
1 -1 <= 0.01
-1 1 <= 0.01
Starting point: 0.9 0.9
Reflective HMC result:
-219.245 -220.487
-354.751 -206.289
-285.74 -101.97
-472.072  99.8196
Step size (final): 1
Discard Ratio: 0.1
Average Acceptance Log-prob: 0
```
This time, volesti at least returns some points, rather than `-nan inf`. However, these points are clearly outside the convex polytope. I didn't have this issue with the R code. Do you happen to know what may explain this difference?

**Comment by LongPham7** [comment]
@papachristoumarios When the negative lop-probability gradient is translated from R to C++, `GradientFunctor` in [this line](https://github.com/GeomScale/volesti/blob/423bbcfc9851b43acd62032c732d633f0d6cb01e/R-proj/src/oracle_functors_rcpp.h#L53) implements `operator()` in the form:
```
if (i == params.order - 1) {
    ....
} else {
    return xs[i + 1]; // returns derivative
}
```
So I believe my original C++ code, rather than your suggested modification, should be equivalent to the R code.

**Comment by LongPham7** [comment]
@vissarion @papachristoumarios I managed to fix the issue. The solution was to compute the inner ball of a convex polytope before performing HMC. So I just need to insert the following code snippet :
```c++
std::pair<Point, NT> InnerBall = P.ComputeInnerBall();
if (InnerBall.second < 0.0)
    throw std::invalid_argument("The linear program is infeasible");
```
I discovered this code snippet [here](https://github.com/GeomScale/volesti/blob/423bbcfc9851b43acd62032c732d633f0d6cb01e/R-proj/src/sample_points.cpp#L615). I have no clue why it solves the problem though - I guess precomputing an inner ball is a prerequisite for HMC. 

Anyway, thanks a lot for your help!
